Sources/SentryCrash/Recording/SentryCrashBinaryImageCache.c

```diff
@@ -0,0 +1,123 @@
+#include <mach-o/dyld.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#define CACHE_SIZE_INCREMENT 100
+static SentryCrashBinaryImage *binaryImagesBuffer = NULL;
+static uint binaryImagesBufferLength;
+static int binaryImagesAmount;
+static pthread_mutex_t binaryImagesMutex = PTHREAD_MUTEX_INITIALIZER;
+
+static void
+increaseBufferSize(void)
+{
+    uint oldLength = binaryImagesBufferLength;
+    uint newLength = oldLength + CACHE_SIZE_INCREMENT;
+
+    SentryCrashBinaryImage *newBuffer = malloc(sizeof(SentryCrashBinaryImage) * newLength);
+    memcpy(newBuffer, binaryImagesBuffer, sizeof(SentryCrashBinaryImage) * oldLength);
+    free(binaryImagesBuffer);
+
+    binaryImagesBuffer = newBuffer;
+    binaryImagesBufferLength = newLength;
+}
+
+static void
+binaryImageAdded(const struct mach_header *header, intptr_t slide)
+{
+    if (binaryImagesBuffer == NULL) {
+        return;
+    }
+
+    Dl_info info;
+    if (!dladdr(header, &info) || info.dli_fname == NULL) {
+        return;
+    }
+
+    SentryCrashBinaryImage binaryImage = { 0 };
+    if (!sentrycrashdl_getBinaryImageForHeader(
+            (const void *)header, info.dli_fname, &binaryImage)) {
+        return;
+    }
+
+    pthread_mutex_lock(&binaryImagesMutex);
+    if (binaryImagesAmount >= binaryImagesBufferLength) {
+        increaseBufferSize();
+    }
+    binaryImagesBuffer[binaryImagesAmount++] = binaryImage;
+    pthread_mutex_unlock(&binaryImagesMutex);
+}
+
+static void
+binaryImageRemoved(const struct mach_header *header, intptr_t slide)
+{
+    pthread_mutex_lock(&binaryImagesMutex);
+    int index = -1;
+    for (int i = 0; i < binaryImagesAmount; i++) {
+        if (binaryImagesBuffer[i].address == (uint64_t)header) {
+            index = i;
+            break;
+        }
+    }
+
+    if (index < 0 || index >= binaryImagesAmount) {
+        pthread_mutex_unlock(&binaryImagesMutex);
+        return;
+    }
+
+    int amountToMove = binaryImagesAmount - index + 1;
+    int sizeToMove = amountToMove * sizeof(SentryCrashBinaryImage);
+    void *startPosition = binaryImagesBuffer + ((index + 1) * sizeof(SentryCrashBinaryImage));
+    void *moveTo = startPosition - sizeof(SentryCrashBinaryImage);
+
+    memcpy(moveTo, startPosition, sizeToMove);
+    binaryImagesAmount--;
+
+    pthread_mutex_unlock(&binaryImagesMutex);
+}
+
+int
+sentrycrashbic_imageCount(void)
+{
+    return binaryImagesAmount;
+}
+
+SentryCrashBinaryImage *
+sentrycrashbic_getBinaryImageCache(int index)
+{
+    if (index >= binaryImagesAmount) {
+        return NULL;
+    }
+    return &binaryImagesBuffer[index];
+}
+
+void
+sentrycrashbic_startCache(void)
+{
+    pthread_mutex_lock(&binaryImagesMutex);
+    binaryImagesBufferLength = CACHE_SIZE_INCREMENT;
+    binaryImagesBuffer = malloc(sizeof(SentryCrashBinaryImage) * binaryImagesBufferLength);
+    binaryImagesAmount = 0;
+    _dyld_register_func_for_add_image(&binaryImageAdded);
```

Comment:
`h`: 

> When this function is first registered it is called for once for each image that is currently part of the process.

After calling `sentrycrashbic_stopCache` and clearing the `binaryImagesBuffer` `_dyld_register_func_for_add_image` will not call our registered function once for each image, so we won't get a full list of images. A couple of ideas on how to solve this problem: 

1. Never stopping updating our cache
2. Doing some check with `_dyld_image_count` and retrieve a full list if `_dyld_image_count` returns a different number.